packages/dev/core/src/Materials/greasedLinePluginMaterial.ts

```diff
@@ -0,0 +1,707 @@
+        const colorTable: Uint8Array = new Uint8Array(colors.length * 4);
+        for (let i = 0, j = 0; i < colors.length; i++) {
+            colorTable[j++] = colors[i].r * 255;
+            colorTable[j++] = colors[i].g * 255;
+            colorTable[j++] = colors[i].b * 255;
+            colorTable[j++] = 255;
+        }
+
+        return colorTable;
+    }
+
+    /**
+     * Creates a RawTexture from an RGBA color array and sets it on the plugin material instance.
+     * @param name name of the texture
+     * @param colors Uint8Array of colors
+     */
+    private _createColorsTexture(name: string, colors: Color3[]) {
+        const colorsArray = GreasedLinePluginMaterial._Color3toRGBAUint8(colors);
+        this._colorsTexture = new RawTexture(
+            colorsArray,
+            colors.length,
+            1,
+            Engine.TEXTUREFORMAT_RGBA,
+            this._scene,
+            false,
+            true,
+            this._options.colorsSampling ?? RawTexture.NEAREST_NEAREST
+        );
+        this._colorsTexture.name = name;
+    }
+
+    /**
+     * Disposes the plugin material.
+     */
+    public dispose(): void {
+        this._colorsTexture?.dispose();
+        super.dispose();
+    }
+
+    /**
+     * Sets whether to use the colors option to colorize the line.
+     * @param value true if use the colors, otherwise false
+     */
+    public setUseColors(value: boolean) {
+        this._options.useColors = value;
+    }
+
+    /**
+     * Creates or updates the colors texture
+     * @param colors color table RGBA
+     * @param lazy if lazy, the colors are not updated
+     * @param forceUpdate force creation of a new texture
+     * @returns
+     */
+    public setColors(colors: Nullable<Color3[]>, lazy = false, forceUpdate = false): void {
+        if (colors === null || colors.length === 0) {
+            this._colorsTexture?.dispose();
+            return;
+        }
+
+        const origColorsCount = this._options.colors?.length ?? 0;
+
+        this._options.colors = colors;
+
+        if (lazy && !forceUpdate) {
+            return;
+        }
+
+        if (this._colorsTexture && origColorsCount === colors.length && !forceUpdate) {
+            const colorArray = GreasedLinePluginMaterial._Color3toRGBAUint8(colors);
+            this._colorsTexture.update(colorArray);
+        } else {
+            this._colorsTexture?.dispose();
+            this._createColorsTexture(`${this._material.name}-colors-texture`, colors);
+        }
+    }
+
+    /**
+     * Updates the material. Use when material created in lazy mode.
+     */
+    public updateLazy() {
+        if (this._options.colors) {
+            this.setColors(this._options.colors, false, true);
+        }
+    }
+
+    /**
+     * Gets the plugin material options
+     * @returns the plugin material options @see GreasedLineMaterialOptions
+     */
+    public getOptions(): GreasedLineMaterialOptions {

```

Comment:
returning options like this (as none readonly) means ppl can change it, which may or may not impact the visual result but without the dirty flags being set when necessary.

The pattern used here with getOptions / set... is also pretty different from what we use in other plugins.

For instance why is color not simply a field of this class or a property to use with getter/setter. Same would go for visibility useDash and so on.

@deltakosh what do you think ?

Reply:
If I remember correctly, we have discussed this with @Popov72 snd agreed that it's not required to clone and freeze the options, it just has to be clearly documented in the docs, that changing the options directly doesn't do anything.

I would lie I'd tall you that I wasn't thinking about this difference. But it's just a matter of approach, like for example some classes in BabylonJS has a static `serialize` method, some of them have instance methods. I used setters in the early version and it remained that way until now.

It'll be absolutely not an issue for me to change to getters/setters - I want GreasedLine to absolutely fit in the BabylonJS ecosystem and be polished to the max ;-) Just let me know, I'll be off at 20:00 today and have computers access on 3rd of July, I can make the changes then.

Reply:
Guys, I can feel that you'll want (logically) the same approach here so I already started to rewrite it to use getters/setters and paused working on the GreasedLine demo ;-)  It might confuse the users if I rewrite the things after two weeks, now that it is already out. Prepare to merge soon :-D